File: check_streaks.js
const fs = require('fs');

const records = JSON.parse(fs.readFileSync('public/data/studyRecords.json', 'utf8'));
const usersData = JSON.parse(fs.readFileSync('public/data/users.json', 'utf8'));
const users = usersData.data;

// Get all users in Challenge 2 (Challenge 9)
const challengeTwo = users.filter(u => u.challengeIds && u.challengeIds.includes(2));
const challengeTwoIds = new Set(challengeTwo.map(u => u.id.toString()));

// Count check-ins for each user
const checkins = {};
Object.keys(records).forEach(date => {
  Object.keys(records[date]).forEach(uid => {
    if (challengeTwoIds.has(uid)) {
      checkins[uid] = (checkins[uid] || 0) + 1;
    }
  });
});

// Find users with 95-100 check-ins
const filtered = Object.entries(checkins)
  .filter(([uid, count]) => count >= 95 && count <= 100)
  .sort((a, b) => b[1] - a[1]);

console.log(`Found ${filtered.length} users with 95-100 check-in days:\n`);

if (filtered.length === 0) {
  console.log('No users found with exactly 95-100 check-in days.');
  console.log('\nTop users by check-in count:');
  Object.entries(checkins)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 15)
    .forEach(([uid, count]) => {
      const user = users.find(u => u.id === parseInt(uid));
      console.log(`ID ${uid}: ${user.name} - ${count} days`);
    });
} else {
  filtered.forEach(([uid, count]) => {
    const user = users.find(u => u.id === parseInt(uid));
    console.log(`ID ${uid}: ${user.name} - ${count} days`);
  });
}

File: check_streaks.py
import json

with open('public/data/studyRecords.json', 'r', encoding='utf-8') as f:
    records = json.load(f)

with open('public/data/users.json', 'r', encoding='utf-8') as f:
    users_data = json.load(f)

users = users_data['data']

# Get users in Challenge 2
challenge_two_ids = {str(u['id']) for u in users if u.get('challengeIds') and 2 in u['challengeIds']}

# Count check-ins
checkins = {}
for date, day_records in records.items():
    for uid, checked in day_records.items():
        if uid in challenge_two_ids and checked:
            checkins[uid] = checkins.get(uid, 0) + 1

# Filter 95-100 days
filtered = [(uid, count) for uid, count in checkins.items() if 95 <= count <= 100]
filtered.sort(key=lambda x: x[1], reverse=True)

if not filtered:
    print('Không có người nào có 95-100 ngày check-in.')
    print()
    print('Top 15 người có số ngày check-in cao nhất:')
    sorted_checkins = sorted(checkins.items(), key=lambda x: x[1], reverse=True)[:15]
    for uid, count in sorted_checkins:
        user = next((u for u in users if u['id'] == int(uid)), None)
        if user:
            print(f"ID {uid}: {user['name']} - {count} ngày")
else:
    print(f'Tìm được {len(filtered)} người có 95-100 ngày check-in:')
    print()
    for uid, count in filtered:
        user = next((u for u in users if u['id'] == int(uid)), None)
        if user:
            print(f"ID {uid}: {user['name']} - {count} ngày")
